src/tools/tidy/src/features.rs

```diff
@@ -110,21 +105,88 @@ pub fn check(path: &Path, bad: &mut bool) {
                 continue;
             }
             lib_features.insert(feature_name.to_owned(),
-                                LibFeature {
+                                Feature {
                                     level: level,
                                     since: since.to_owned(),
+                                    has_gate_test: false,
                                 });
         }
     });
 
+    super::walk(&path.join("test/compile-fail"),
+                &mut |path| super::filter_dirs(path),
+                &mut |file| {
+        let filename = file.file_name().unwrap().to_string_lossy();
+        if !filename.ends_with(".rs") || filename == "features.rs" ||
+           filename == "diagnostic_list.rs" {
+            return;
+        }
+
+        contents.truncate(0);
+        t!(t!(File::open(&file), &file).read_to_string(&mut contents));
+
+        for (i, line) in contents.lines().enumerate() {
+            let mut err = |msg: &str| {
+                println!("{}:{}: {}", file.display(), i + 1, msg);
+                *bad = true;
+            };
+
+            let gate_test_str = "gate-test-";
+
+            if !line.contains(gate_test_str) {
+                continue;
+            }
+
+            let feature_name = match line.find(gate_test_str) {
+                Some(i) => {
+                    &line[i+gate_test_str.len()..line[i+1..].find(' ').unwrap_or(line.len())]
+                },
+                None => continue,
+            };
+            let found_feature = features.get_mut(feature_name)
+                                        .map(|v| { v.has_gate_test = true; () })
+                                        .is_some();
+
+            let found_lib_feature = features.get_mut(feature_name)
+                                            .map(|v| { v.has_gate_test = true; () })
+                                            .is_some();
+
+            if !(found_feature || found_lib_feature) {
+                err(&format!("gate-test test found referencing a nonexistent feature '{}'",
+                             feature_name));
+            }
+        }
+    });
+
+    // Only check the number of lang features.
+    // Obligatory testing for library features is dumb.
+    let gate_untested = features.iter()
+                                .filter(|&(_, f)| f.level == Status::Unstable)
+                                .filter(|&(_, f)| !f.has_gate_test)
+                                .count();
+
+    // FIXME get this number down to zero.
+    let gate_untested_expected = 94;
```

Comment:
Sure, good idea! Most of the features already have gate tests (see issue #22820 for an outdated list). If the general direction of this issue is agreed upon, I volunteer to add `// gate-test-` comments to the already existing tests (that I can find), and for the small remaining number we can open a tracking issue. Probably a good thing to ask contributions for from first time contributors (like the error message refactoring).